client/file-manager/src/in_memory.rs

```diff
@@ -1,40 +1,167 @@
+
+        let chunk: Option<Vec<u8>> = trie
+            .get(&chunk_id.to_be_bytes())
+            .map_err(|_| FileStorageError::FailedToGetFileChunk)?;
+
+        let chunk = chunk.ok_or(FileStorageError::FileChunkDoesNotExist)?;
+
+        // Drop the `trie_recorder` to release the `recorder`
+        drop(trie_recorder);
+
+        // Generate proof
+        let proof = recorder
+            .drain_storage_proof()
+            .to_compact_proof::<T::Hash>(file_data.root)
+            .map_err(|_| FileStorageError::FailedToGenerateCompactProof)?;
+
+        Ok(FileProof {
+            proven: Leaf {
+                key: (*chunk_id).clone(),
+                data: chunk,
+            },
+            proof,
+            root: file_data.get_root(),
+        })
     }
 
-    fn delete_file(&self, _key: &Self::Key) {
-        unimplemented!()
+    fn delete_file(&mut self, file_key: &Key) {
+        self.metadata.remove(file_key);
+        self.file_data.remove(file_key);
     }
 
-    fn get_metadata(&self, _key: &Self::Key) -> Option<Metadata> {
-        unimplemented!()
+    fn get_metadata(&self, file_key: &Key) -> Result<Metadata, FileStorageError> {
+        self.metadata
+            .get(file_key)
+            .cloned()
+            .ok_or(FileStorageError::FileDoesNotExist)
     }
 
-    fn set_metadata(&self, _key: &Self::Key, _metadata: &Metadata) {
-        unimplemented!()
+    fn set_metadata(&mut self, file_key: Key, metadata: Metadata) {
+        self.metadata.insert(file_key, metadata);
+        self.file_data.insert(file_key, FileData::new());
     }
 
-    fn get_chunk(&self, _key: &Self::Key) -> Option<Self::Value> {
-        unimplemented!()
+    fn get_chunk(&self, file_key: &Key, chunk_id: &ChunkId) -> Result<Chunk, FileStorageError> {
+        let file_data = self.file_data.get(file_key);
+        let file_data = file_data.ok_or(FileStorageError::FileDoesNotExist)?;
+
+        let trie = TrieDBBuilder::<T>::new(&file_data.memdb, &file_data.root).build();
+
+        Ok(trie
+            .get(&chunk_id.to_be_bytes())
+            .map_err(|_| FileStorageError::FailedToGetFileChunk)?
+            .ok_or(FileStorageError::FileChunkDoesNotExist)?)
     }
 
-    fn write_chunk(&self, _key: &str, _data: &Self::Value) {
-        unimplemented!()
+    fn write_chunk(
+        &mut self,
+        file_key: &Key,
+        chunk_id: &ChunkId,
+        data: &Chunk,
+    ) -> Result<(), FileStorageError> {
+        let file_data = self
+            .file_data
+            .get_mut(file_key)
+            .ok_or(FileStorageError::FileDoesNotExist)?;
+
+        let mut trie =
+            TrieDBMutBuilder::<T>::new(&mut file_data.memdb, &mut file_data.root).build();
+
+        // Check that we don't have a chunk already stored.
+        if trie
+            .contains(&chunk_id.to_be_bytes())
+            .map_err(|_| FileStorageError::FailedToGetFileChunk)?
+        {
+            return Err(FileStorageError::FileChunkAlreadyExists);
```

Comment:
So this doesn't allow you to re-write a chunk? I guess for our scope no it's ok like this though.